lib/runner.go

```diff
@@ -26,6 +26,36 @@ import (
 	"github.com/loadimpact/k6/stats"
 )
 
+// ActiveVU represents an actively running virtual user.
+type ActiveVU interface {
+	// Runs the VU once. The only way to interrupt the execution is to cancel
+	// the context given to InitializedVU.Activate()
+	RunOnce() error
+}
+
+// InitializedVU represents a virtual user ready for work. It needs to be
+// activated (i.e. given a context) before it can actually be used. Activation
+// also requires a callback function, which will be called when the supplied
+// context is done. That way, VUs can be returned to a pool and reused.
+type InitializedVU interface {
+	// Fully activate the VU so it will be able to run code
+	Activate(*VUActivationParams) ActiveVU
+}
+
+// VUActivationParams are supplied by each executor when it retrieves a VU from
+// the buffer pool and activates it for use.
+type VUActivationParams struct {
+	RunContext         context.Context
+	DeactivateCallback func()
+	// Used to signal when the VU has finished execution, after RunContext
+	// is done, the Goja VM is interrupted and the callback is called.
+	// It shouldn't be closed.
+	Done chan struct{}
```

Comment:
Hhmm yes, I suppose it could be handled in the `DeactivateCallback` as well, didn't think of that even though a `WaitGroup` was used in one of the tests.

OK, I'll refactor this, thanks.

Reply:
Resolved in the latest push.